From: ROCKNIX Contributors <rocknix@rocknix.org>
Date: Mon, 28 Apr 2026 00:00:00 +0000
Subject: [PATCH] arm64: dts: qcom: cq8725s-ayn-odin3: Add fan cooling maps for CPU and GPU clusters

Wire the pwm-fan into the thermal framework for all four CPU performance
clusters (cpuss-0-0, cpuss-0-1, cpuss-1-0, cpuss-1-1) and all eight GPU
clusters (gpuss0 through gpuss7). All changes are isolated to the Odin 3
device tree so that future SM8750 devices are unaffected.

Fan states (8 levels): 0=off, 1-7=30/45/60/70/90/120/150 PWM
Curve: 40C->L1, 50C->L2, 60C->L3, 65C->L4, 70C->L5, 75C->L6, 80C->L7
Hysteresis: 3C on all active trips

Depends-on: 0601-ROCKNIX-sm8750-thermal-zones.patch
---
diff --git a/arch/arm64/boot/dts/qcom/cq8725s-ayn-odin3.dts b/arch/arm64/boot/dts/qcom/cq8725s-ayn-odin3.dts
--- a/arch/arm64/boot/dts/qcom/cq8725s-ayn-odin3.dts
+++ b/arch/arm64/boot/dts/qcom/cq8725s-ayn-odin3.dts
@@ -311,3 +311,272 @@
 		drive-strength = <2>;
 	};
 };
+
+&pwm_fan {
+	#cooling-cells = <2>;
+	cooling-levels = <0 30 45 60 70 90 120 150>;
+};
+
+&cpuss_0_0_thermal {
+	trips {
+		cpu00_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		cpu00_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		cpu00_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		cpu00_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		cpu00_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		cpu00_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		cpu00_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&cpu00_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&cpu00_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&cpu00_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&cpu00_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&cpu00_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&cpu00_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&cpu00_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&cpuss_0_1_thermal {
+	trips {
+		cpu01_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		cpu01_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		cpu01_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		cpu01_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		cpu01_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		cpu01_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		cpu01_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&cpu01_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&cpu01_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&cpu01_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&cpu01_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&cpu01_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&cpu01_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&cpu01_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&cpuss_1_0_thermal {
+	trips {
+		cpu10_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		cpu10_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		cpu10_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		cpu10_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		cpu10_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		cpu10_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		cpu10_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&cpu10_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&cpu10_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&cpu10_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&cpu10_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&cpu10_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&cpu10_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&cpu10_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&cpuss_1_1_thermal {
+	trips {
+		cpu11_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		cpu11_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		cpu11_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		cpu11_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		cpu11_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		cpu11_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		cpu11_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&cpu11_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&cpu11_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&cpu11_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&cpu11_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&cpu11_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&cpu11_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&cpu11_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&gpuss0_thermal {
+	trips {
+		gpu0_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		gpu0_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		gpu0_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		gpu0_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		gpu0_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		gpu0_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		gpu0_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&gpu0_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&gpu0_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&gpu0_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&gpu0_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&gpu0_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&gpu0_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&gpu0_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&gpuss1_thermal {
+	trips {
+		gpu1_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		gpu1_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		gpu1_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		gpu1_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		gpu1_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		gpu1_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		gpu1_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&gpu1_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&gpu1_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&gpu1_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&gpu1_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&gpu1_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&gpu1_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&gpu1_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&gpuss2_thermal {
+	trips {
+		gpu2_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		gpu2_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		gpu2_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		gpu2_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		gpu2_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		gpu2_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		gpu2_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&gpu2_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&gpu2_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&gpu2_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&gpu2_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&gpu2_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&gpu2_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&gpu2_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&gpuss3_thermal {
+	trips {
+		gpu3_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		gpu3_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		gpu3_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		gpu3_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		gpu3_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		gpu3_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		gpu3_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&gpu3_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&gpu3_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&gpu3_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&gpu3_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&gpu3_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&gpu3_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&gpu3_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&gpuss4_thermal {
+	trips {
+		gpu4_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		gpu4_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		gpu4_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		gpu4_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		gpu4_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		gpu4_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		gpu4_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&gpu4_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&gpu4_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&gpu4_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&gpu4_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&gpu4_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&gpu4_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&gpu4_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&gpuss5_thermal {
+	trips {
+		gpu5_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		gpu5_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		gpu5_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		gpu5_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		gpu5_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		gpu5_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		gpu5_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&gpu5_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&gpu5_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&gpu5_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&gpu5_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&gpu5_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&gpu5_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&gpu5_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&gpuss6_thermal {
+	trips {
+		gpu6_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		gpu6_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		gpu6_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		gpu6_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		gpu6_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		gpu6_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		gpu6_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&gpu6_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&gpu6_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&gpu6_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&gpu6_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&gpu6_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&gpu6_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&gpu6_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
+
+&gpuss7_thermal {
+	trips {
+		gpu7_fan0: trip-point-fan0 { temperature = <40000>; hysteresis = <3000>; type = "active"; };
+		gpu7_fan1: trip-point-fan1 { temperature = <50000>; hysteresis = <3000>; type = "active"; };
+		gpu7_fan2: trip-point-fan2 { temperature = <60000>; hysteresis = <3000>; type = "active"; };
+		gpu7_fan3: trip-point-fan3 { temperature = <65000>; hysteresis = <3000>; type = "active"; };
+		gpu7_fan4: trip-point-fan4 { temperature = <70000>; hysteresis = <3000>; type = "active"; };
+		gpu7_fan5: trip-point-fan5 { temperature = <75000>; hysteresis = <3000>; type = "active"; };
+		gpu7_fan6: trip-point-fan6 { temperature = <80000>; hysteresis = <3000>; type = "active"; };
+	};
+
+	cooling-maps {
+		map0 { trip = <&gpu7_fan0>; cooling-device = <&pwm_fan 0 1>; };
+		map1 { trip = <&gpu7_fan1>; cooling-device = <&pwm_fan 1 2>; };
+		map2 { trip = <&gpu7_fan2>; cooling-device = <&pwm_fan 2 3>; };
+		map3 { trip = <&gpu7_fan3>; cooling-device = <&pwm_fan 3 4>; };
+		map4 { trip = <&gpu7_fan4>; cooling-device = <&pwm_fan 4 5>; };
+		map5 { trip = <&gpu7_fan5>; cooling-device = <&pwm_fan 5 6>; };
+		map6 { trip = <&gpu7_fan6>; cooling-device = <&pwm_fan 6 7>; };
+	};
+};
